source/math/Quaternion.ooc

```diff
@@ -99,17 +187,39 @@ Quaternion: cover {
 		this y ==  y &&
 		this z == other z
 	}
+	/*operator + (other: This) -> This {
+		This new (this w + other w, this x + other x, this y + other y, this z + other z)
+	}*/
 	operator + (other: This) -> This {
-		This new(this w + other w, this x + other x, this y + other y, this z + other z)
+		This new (this real + other real, this imaginary + other imaginary)
 	}
+	/*operator - (other: This) -> This {
+		This new (this w - other w, this x - other x, this y - other y, this z - other z)
+	}*/
 	operator - (other: This) -> This {
-		This new(this w - other w, this x - other x, this y - other y, this z - other z)
+		this + (-other)
+	}
+	operator - -> This {
+		This new (-this real, -this imaginary)
 	}
 	operator / (value: Float) -> This {
-		This new(this w / value, this x / value, this y / value, this z / value)
+		This new (this w / value, this x / value, this y / value, this z / value)
 	}
 	operator * (value: Float) -> This {
-		This new(this w * value, this x * value, this y * value, this z * value)
+		This new (this w * value, this x * value, this y * value, this z * value)
+	}
+	operator * (other: This) -> This {
+		realResult := this real * other real - this imaginary scalarProduct(other imaginary)
+		imaginaryResult := this real * other imaginary + this imaginary * other real + this imaginary vectorProduct(other imaginary)
+		This new (realResult, imaginaryResult)
 	}
-	toString: func -> String { "Real:" + "%8f" format(this real) + " Imaginary: " + "%8f" format(this imaginary x) + " " + "%8f" format(this imaginary y) + " " + "%8f" format(this imaginary z)}
+	operator * (value: FloatPoint3D) -> FloatPoint3D {
+		(this * Quaternion new (0.0f, value) * this inverse) imaginary
+	}
+	toString: func -> String {
+		"Real:" + "%8f" format(this real) +
+		" Imaginary: " + "%8f" format(this imaginary x) + " " + "%8f" format(this imaginary y) + " " + "%8f" format(this imaginary z)}
```

Comment:
This is going to leak memory.

```ooc
"Real: " << "%8f" formatFloat(this real) >>
" Imaginary: " & "%8f" formatFloat(this imaginary x) >> " " & "%8f" formatFloat(this imaginary y) >> " " & "%8f" formatFloat(this imaginary z)
```
should do the trick. The operators aren't pretty, but they work well when you get them right. For reference,
`<<` keeps the left argument, frees the right one, and returns the concatenation of both arguments
`>>` keeps the right argument, frees the left one, and returns the concatenation of both arguments
`&` frees both arguments and returns the concatenation of both arguments
`+` keeps both arguments and returns the concatenation of both arguments